src/protocols/interaction_model/StatusCode.h

```diff
@@ -48,6 +50,108 @@ enum class Status : uint8_t
 const char * StatusName(Status status);
 #endif // CHIP_CONFIG_IM_STATUS_CODE_VERBOSE_FORMAT
 
+/**
+ * @brief Class to encapsulate a Status code, including possibly a
+ *        cluster-specific code for generic SUCCESS/FAILURE.
+ *
+ * This can be used everywhere a `Status` is used, but it is lossy
+ * to the cluster-specific code if used in place of `Status` when
+ * the cluster-specific code is set.
+ *
+ * This class can only be directly constructed from a `Status`. To
+ * attach a cluster-specific-code, please use the `ClusterSpecificFailure`
+ * and `ClusterSpecificSuccess` factory methods.
+ */
+class StatusCode
+{
+public:
+    explicit StatusCode(Status status) : mStatus(status) {}
+
+    // We only have simple copyable members, so we should be trivially copyable.
+    StatusCode(const StatusCode & other)             = default;
+    StatusCode & operator=(const StatusCode & other) = default;
+
+    bool operator==(const StatusCode & other)
+    {
+        return (this->mStatus == other.mStatus) && (this->HasClusterSpecificCode() == other.HasClusterSpecificCode()) &&
+            (this->GetClusterSpecificCode() == other.GetClusterSpecificCode());
+    }
+
+    bool operator!=(const StatusCode & other) { return !(*this == other); }
+
+    StatusCode & operator=(const Status & status)
+    {
+        this->mStatus              = status;
+        this->mClusterSpecificCode = chip::NullOptional;
+        return *this;
+    }
+
+    /**
+     * @brief Builder for a cluster-specific failure status code.
+     *
+     * @tparam T - enum type for the cluster-specific status code
+     *             (e.g. chip::app::Clusters::AdministratorCommissioning::CommissioningWindowStatusEnum)
+     * @param cluster_specific_code - cluster-specific code to record with the failure
+     *             (e.g. chip::app::Clusters::AdministratorCommissioning::CommissioningWindowStatusEnum::kWindowNotOpen)
+     * @return a StatusCode instance properly configured.
+     */
+    template <typename T>
+    static StatusCode ClusterSpecificFailure(T cluster_specific_code)
+    {
+        static_assert(std::numeric_limits<T>::max() <= std::numeric_limits<uint8_t>::max(), "Type used must fit in uint8_t");
+        return StatusCode(Status::Failure, static_cast<uint8_t>(cluster_specific_code));
```

Comment:
If this used `to_underlying` instead of explicit cast, we would not need the static_assert (given that we use -Wconversion), yes?

Reply:
I guess I'm not 100% sure what would fail with -Wconversion (i.e. I don't like being a type system in my head). This makes it explicit. I can replace to `to_underlying` but I'm having a hard time reasoning about if there are corner cases of enums being used that would fail. I guess if they are all `enum BlahBlah : uint8_t`, they wouldn't, and otherwise would.

Reply:
Right, that's the idea.  You use `to_underlying` to get the actual underlying type, and then calling the StatusCode constructor fails if that type is not `uint8_t` because of `-Wconversion`.